### PR TITLE
MRG: Concluding remarks

### LINKED ISSUE #261: Conclusion / Final wrap-up
One short page:
- pointer to scikit-learn doc
- how the doc is organised
  + user guide
  + API
  + examples
- doc scikit-learn
- where to ask questions

### PR DESCRIPTION
**Pull request by GaelVaroquaux**
Closes #261.

This now has the text corresponding to the video and a little more that we had removed to make the video shorter.

### CODE REVIEW

**Comment by GaelVaroquaux** [comment]
This is now ready for reviews and merge.

File: jupyter-book/concluding_remarks.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+### 3. Specific models
+
+- Understanding the models
+  - Helps knowing when they are suited to the data 
+  - Gives intuitions on how to debug them
+
+- Linear models
+  - build predictions by combining the values of features
+  - Particularly useful for data with many features or few observations
+
```
Comment (ogrisel) [inline jupyter-book/concluding_remarks.md:60]:
```suggestion
  - Particularly useful for data with many features or few observations
  - Can benefit from non-linear feature engineering
```

File: jupyter-book/concluding_remarks.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+- The [scikit-learn doc](http://scikit-learn.org)
+    - The documentation is rich, didactic, continuously improving
+- These docs comprise
+  - An user guide: Gives the intuition behind every machine-learning
+    method, and how it can be useful
+  - API docs: Every function, every parameter is explained
+  - Examples: each example tries to demonstrate the good use of the software
+- Where to ask questions:
+    - Stackoverflow
```
Comment (ogrisel) [inline jupyter-book/concluding_remarks.md:85]:
```suggestion
    - [Stackoverflow](https://stackoverflow.com/)
    - [GitHub discussions for scikit-learn](https://github.com/scikit-learn/scikit-learn/discussions)
```

File: jupyter-book/concluding_remarks.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+
+- Model inspection
+    - Understanding what drives a prediction
+    - Useful for debuging, for reasonning on the system at hand
+    - Requires a lot of nuance
+
+- Deep learning
+    - For images, text, voice: use pretrained models
+    - Comes with great computational and human costs, as well as large maintenance costs
+    - Not in scikit-learn
```
Comment (ogrisel) [inline jupyter-book/concluding_remarks.md:113]:
```suggestion
    - Not in scikit-learn: have a look at resources on pytorch and tensorflow to get started!
```

File: jupyter-book/concluding_remarks.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    - Finding order and structure in the data, for instance to group samples, or to transform features
+    - Particularly useful because it does not need labels
+    - But given labels, supervised learning not unsupervised learning, is more likely to recover the link
+
+- Model inspection
+    - Understanding what drives a prediction
+    - Useful for debuging, for reasonning on the system at hand
+    - Requires a lot of nuance
+
+- Deep learning
```
Comment (ogrisel) [inline jupyter-book/concluding_remarks.md:110]:
```suggestion
- Deep learning
    - Often not better than gradient boosting trees for classification or regression
      on tabular data
    - But more flexible: can work natively with tasks that involve variable length
      structures in the input and output of the model (e.g. speech to text)
```

File: jupyter-book/concluding_remarks.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+
+- [Introduction to Machine Learning with
+  Python](https://www.oreilly.com/library/view/introduction-to-machine/9781449369880/)
+  by Andreas C. Müller, Sarah Guido: explains more advanced use of
+  scikit-learn
+- [Python Data Science Handbook](https://jakevdp.github.io/PythonDataScienceHandbook/) by Jake van der Plas, a broader picture of data science,
+  beyond scikit-learn
+- [An Introduction to Statistical Learning](https://www.statlearning.com/),
+  by James, Witten, Hastie, Tibshirani: statistical theory 
+  behind the concepts that we have explored
```
Comment (ogrisel) [inline jupyter-book/concluding_remarks.md:125]:
```suggestion
  behind the concepts that we have explored
- [Hands-On Machine Learning with Scikit-Learn, Keras, and TensorFlow](https://www.oreilly.com/library/view/hands-on-machine-learning/9781492032632/) by Aurélien Géron
```

**Review by ogrisel — COMMENTED** [review]
Partial feedback, so far it looks good!

**Comment by GaelVaroquaux** [comment]
Thanks for the partial review and the great changes,  @ogrisel !

File: jupyter-book/concluding_remarks.md [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+- Transformations of the data are often necessary
+  - Typically, encoding of the categorical variables
+  - They must only use information available at train time
+  - For this, use the scikit-learn Pipeline object
+
+### 2. Adapting model complexity to the data
+
+- Models seek to minimize the error on the test set
+  - Minimizing error on the train set does no suffice
```
Comment (ArturoAmorQ) [inline jupyter-book/concluding_remarks.md:41]:
```suggestion
  - Minimizing error on the train set does not suffice
```

File: jupyter-book/concluding_remarks.md [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+    - Contributing code is technical
+	- Learn software engineering:
+	- Learn git, github (https://lab.github.com/)
+
+### Topics we have not covered
+
+- Unsupervised learning
+    - Finding order and structure in the data, for instance to group samples, or to transform features
+    - Particularly useful because it does not need labels
+    - But given labels, supervised learning not unsupervised learning, is more likely to recover the link
```
Comment (ArturoAmorQ) [inline jupyter-book/concluding_remarks.md:104]: I think this line is not clear enough, which link does it refer to?

File: jupyter-book/concluding_remarks.md [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+- Model inspection
+    - Understanding what drives a prediction
+    - Useful for debuging, for reasonning on the system at hand
```
Comment (ArturoAmorQ) [inline jupyter-book/concluding_remarks.md:108]:
```suggestion
    - Useful for debugging, for reasoning on the system at hand
```

File: jupyter-book/concluding_remarks.md [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    - Advertisement: errors are often harmless
+    - Medicine: errors can kill
+- Operational logic: Better a false detection or a miss?
+    - Detecting brain tumors:
+        - If a patient is sent to surgery: false detections are very dangerous
+        - If a patient given an MR scan to confirm the detection:
+	  misses should be avoided, as an MR scan is harmless
+
+The predictions may modify how the system is functions:
+  - Predicting who will benefit from an hospital stay may overcrowd some units of the hospital, and thus change the positive impact of hospitals on inpatients
```
Comment (ArturoAmorQ) [inline jupyter-book/concluding_remarks.md:191]:
```suggestion
The predictions may modify how the system functions:
  - Predicting who will benefit from a hospital stay may overcrowd some units of the hospital, and thus change the positive impact of hospitals on inpatients
```

File: jupyter-book/concluding_remarks.md [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    - It may change with time, it may be uneven across the population (eg higher quality data for rich people)
+- The state of affaires may not be the desired one
+    - For equal qualifications and responsibilities, women are typically payed less than men. A learner will pick this up and amplify inequalities
+
+### Prediction models versus causal models
+
+Machine learner models are not driven by causal mechanisms.
+
+- For example people that go to the hospital die more than people who do not:
+    - Naive data analysis might conclude that hospital are bad for health
```
Comment (ArturoAmorQ) [inline jupyter-book/concluding_remarks.md:216]:
```suggestion
    - Naive data analysis might conclude that hospitals are bad for the health
```

File: jupyter-book/concluding_remarks.md [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+      people who do not.
+- Another example: having a heart pressure greater than a threshold
+  may trigger specific care which is good. A learner will pick up
+  above-threshold heart pressure as predictor of a health improvement
+- In pure predictive settings, these informations are beneficial for the
+  predictions. However:
+ - they should not be trusted when designing interventions
+ - predictive models built on such non-causal information may be brittle
+   to changes of operational procedures
+ - interpretation is subject to caution
```
Comment (ArturoAmorQ) [inline jupyter-book/concluding_remarks.md:228]:
```suggestion
 - interpretation must be subject to caution
```

File: jupyter-book/concluding_remarks.md [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+  - Responsible use of machine learning involves challenges at the
+    intersection of technology and society. No solution will be
+    purely technical
+
+----
+
+**Your move: choose what you will do with machine learning**
+
+
+- Machine learning drives one of the most important technological revolution of our time.
```
Comment (ArturoAmorQ) [inline jupyter-book/concluding_remarks.md:256]:
```suggestion
- Machine learning drives one of the most important technological revolutions of our time.
```

File: jupyter-book/concluding_remarks.md [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+- It is a fantastic opportunity to improve human condition
+- With scikit-learn, and this MOOC, we try to lift as much as possible
+  the technical roadblocks, and  wee hope that we can empower a great variety of people, with different mindsets and dreams, to solve the problems that matters to them
```
Comment (ArturoAmorQ) [inline jupyter-book/concluding_remarks.md:259]:
```suggestion
  the technical roadblocks, and  wee hope that we can empower a great variety of people, with different mindsets and dreams, to solve the problems that matter to them
```

File: slides/concluding_remarks.md [hunk 14]
```diff
@@ -0,0 +1,359 @@
...
+- Scikit-learn: lifting technical roadblocks as much as possible
+    - empower people
+    - to solve the problems that matters to them
+
+???
+
+- Machine learning drives one of the most important technological revolution of our time.
+- It is a fantastic opportunity to improve human condition
+- With scikit-learn, and this MOOC, we try to lift as much as possible
+  the technical roadblocks, and  wee hope that we can empower a great variety of people, with different mindsets and dreams, to solve the problems that matters to them
```
Comment (ArturoAmorQ) [inline slides/concluding_remarks.md:357]:
```suggestion
  the technical roadblocks, and  wee hope that we can empower a great variety of people, with different mindsets and dreams, to solve the problems that matter to them
```

File: slides/concluding_remarks.md [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+.center[
+# Your move
+]
+
+
+- Machine learning drives one of the most important technological revolution of our time.
+- It is a fantastic opportunity to improve our world
+- Scikit-learn: lifting technical roadblocks as much as possible
+    - empower people
+    - to solve the problems that matters to them
```
Comment (ArturoAmorQ) [inline slides/concluding_remarks.md:350]:
```suggestion
    - to solve the problems that matter to them
```

File: slides/concluding_remarks.md [hunk 16]
```diff
@@ -0,0 +1,359 @@
...
+
+???
+
+- Machine learning drives one of the most important technological revolution of our time.
```
Comment (ArturoAmorQ) [inline slides/concluding_remarks.md:354]:
```suggestion
- Machine learning drives one of the most important technological revolutions of our time.
```

**Review by ArturoAmorQ — COMMENTED** [review]
I made some minor corrections that preserve the structure of the content as the video is already recorded. 

Do you think it is worth placing the video in the jupyterhub? If so, maybe we could add such placement in this PR as a TO DO in a md file similar to what was done in  #497.

**Comment by MarieCollin** [comment]
Here is the suggestion for the Conclusion module structure in FUN with two parts:
1- Concluding remarks with video (+slides if needed) and md page for complementary ressources
2- Your general feedback:  on the forum and/or via the anonymous Fun survey
Capture d’écran 2022-01-25 à 14 12 57

What do you think ?

**Comment by GaelVaroquaux** [comment]
> What do you think ?

I think that is a great vision.

How about: we merge this PR first, and implement this vision on top?

**Comment by GaelVaroquaux** [comment]
Thanks for all your fixed, @ArturoAmorQ !

File: jupyter-book/concluding_remarks.md (hunk 13 above)
Comment (ArturoAmorQ) [reply jupyter-book/concluding_remarks.md]: In this case _to matter_ is the verb and its conjugation for the 3rd person in plural is "the problems (they) matter".

https://www.quora.com/When-should-I-use-matter-or-matters-As-in-if-I-write-she-matters-to-me-Will-it-be-correct

The same applies to the other comments I made.

File: jupyter-book/concluding_remarks.md (hunk 12 above)
Comment (ArturoAmorQ) [reply jupyter-book/concluding_remarks.md]: I think "revolutions" here should be in plural as it is meant to be _one out of several_ important technological revolutions

**Comment by GaelVaroquaux** [comment]
I've made one last change. I believe that this is ready to merge.

**Comment by ArturoAmorQ** [comment]
> I believe that this is ready to merge.

Some of my comments have not yet been addressed.

**Comment by ArturoAmorQ** [comment]
> > Some of my comments have not yet been addressed.
> 
> Ooops, I'm sorry. Which one?

Hidden behind the "10 hidden conversations Load more" button, but you can visualize them more easily in the "Files changed". Worry not, they are mostly typos and a comment of one line I didn't quite understood.

File: jupyter-book/concluding_remarks.md (hunk 11 above)
Comment (GaelVaroquaux) [reply jupyter-book/concluding_remarks.md:228]: I think that "is subject to" is preferable

File: jupyter-book/concluding_remarks.md (hunk 7 above)
Comment (GaelVaroquaux) [reply jupyter-book/concluding_remarks.md]: OK, I reformulated.

**Comment by GaelVaroquaux** [comment]
Indeed, I had not seen them. Thanks, I made the modifications

File: jupyter-book/concluding_remarks.md (hunk 11 above)
Comment (ArturoAmorQ) [reply jupyter-book/concluding_remarks.md:228]: Alright, not a strong opinion on this one, I thought we could use "must to" with the intention to "strongly recommend"

**Review by ArturoAmorQ — APPROVED** [review]
LGTM

File: jupyter-book/concluding_remarks.md (hunk 11 above)
Comment (ogrisel) [reply jupyter-book/concluding_remarks.md:228]: Maybe we could have some kind of conclusion to this section:

Quantifying the expected causal effect of a real world intervention is often best
achieved by randomized control trials or A/B tests. In comparison, purely
data-driven observational studies are often tricky to get right in practice.
If you want to learn more about this topic, you might want to explore the
following resources:

- https://github.com/microsoft/EconML
- https://github.com/uber/causalml

File: jupyter-book/concluding_remarks.md [hunk 17]
```diff
@@ -0,0 +1,265 @@
...
+ - interpretation is subject to caution
+
+
+### Societal impact
+
+These challenges with biases in the data, feedback loops of the
+predictions, can be very important, because **prediction models may affect
+people's lives**.
+
+Today, AI systems can be used to allocate loans, screen job applicants,
```
Comment (ogrisel) [inline jupyter-book/concluding_remarks.md:237]:
```suggestion
Today, AI systems are sometimes used to allocate loans, screen job applicants,
```

File: jupyter-book/concluding_remarks.md [hunk 18]
```diff
@@ -0,0 +1,265 @@
...
+prioritise medical treatement, help law enforcement or court decisions.
+
+If you know scikit-learn, [fairlearn](https://fairlearn.org) is a simple
+resource to understand some problems caused by a too naive application of
+machine learning methods.
```
Comment (ogrisel) [inline jupyter-book/concluding_remarks.md:242]:
```suggestion
resource to help you understand, assess and sometimes mitigate some
problems caused by a too naive application of machine learning
methods.
```

File: jupyter-book/concluding_remarks.md [hunk 19]
```diff
@@ -0,0 +1,265 @@
...
+
+----
+
+**Your move: choose what you will do with machine learning**
+
+
+- Machine learning drives one of the most important technological revolutions of our time.
+- It is a fantastic opportunity to improve human condition
+- With scikit-learn, and this MOOC, we try to lift as much as possible
+  the technical roadblocks, and  wee hope that we can empower a great variety of people, with different mindsets and dreams, to solve the problems that matter to them
```
Comment (ogrisel) [inline jupyter-book/concluding_remarks.md:263]:
```suggestion
- With scikit-learn, and this MOOC, we try to lift as many technical roadblocks
  as possible, and  we hope that we can empower a great variety of people, with different mindsets and dreams, to solve the problems that matter to them
```

**Review by ogrisel — APPROVED** [review]
More feedback, but other than that, LGTM.

Feel free to ignore things you do not agree with.

File: jupyter-book/concluding_remarks.md [hunk 20]
```diff
@@ -0,0 +1,265 @@
...
+Errors mean different things in different application contexts.
+
+- Operational risk:
+    - Advertisement: errors are often harmless
+    - Medicine: errors can kill
+- Operational logic: Better a false detection or a miss?
+    - Detecting brain tumors:
+        - If a patient is sent to surgery: false detections are very dangerous
+        - If a patient given an MR scan to confirm the detection:
+	  misses should be avoided, as an MR scan is harmless
```
[the remark this answers is not in the document]
Comment (GaelVaroquaux) [reply jupyter-book/concluding_remarks.md:199]: Fixed

**Comment by GaelVaroquaux** [comment]
I have addressed most of the comments, though I chose to leave out a couple for editorial reasons.

This is now ready to merge, thanks!

**Pull request merged**